torch/testing/_internal/common_methods_invocations.py

```diff
@@ -477,7 +479,24 @@ def sample_inputs(self, device, dtype, requires_grad=False):
 ]
 
 if TEST_SCIPY:
+    def reference_sigmoid(x):
+        # 'scipy.special.expit' not supported for the input types
+        if x.dtype in [np.complex64, np.complex128]:
+            return (1 / (1 + np.exp(-x)))
+        return scipy.special.expit(x)
+
     op_db_scipy_reference = [
+        UnaryUfuncInfo('sigmoid',
+                       ref=reference_sigmoid,
```

Comment:
I think this test can be done all the time with the NumPy based reference and then when SciPy is available it can use scipy.special.expit for non-complex values. That will ensure better test coverage.

The easiest way to do this is probably to add it to the regular op_dbs with a reference that tests for TEST_SCIPY internally.

Reply:
The problem with reference implementation is
```python
import torch
import numpy as np
import scipy.special

def reference_sigmoid(x):
    return 1 / (1 + np.exp(-x))

x = torch.tensor(1, dtype=torch.uint8)

print(torch.sigmoid(x))
print(scipy.special.expit(x.numpy()))
print(reference_sigmoid(x.numpy()))
```

Output:
```
tensor(0.7311) # torch.sigmoid
0.7310586 # scipy.special.expit
test/test_exp.py:6: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-x))
0.0 # Reference
```

The reason this happens is
```python
>>> x = torch.tensor(1, dtype=torch.uint8)
>>> x.numpy()
array(1, dtype=uint8)
>>> -(x.numpy())
255 # hence the overflow encountered in exp above
```

I believe it is better to use scipy implementation only in general.

Reply:
That makes sense. Thank you for explaining it so clearly.

We might be able to work around this with upcasting uint8 inputs, but this seems fine as-is.